Add h3ToCenterChild and getPentagonsAtRes

zachasme opened the pull request:
Resolves #255 and #263.

zachasme commented:
What are your thoughts on naming?

The `atRes` suffix is not used anywhere else, perhaps `getPentagonalIndexes`/`pentgonalIndexCount` would be better?

nrabinowitz commented on src/apps/testapps/testH3ToCenterChild.c:
```diff
@@ -0,0 +1,47 @@
+#include "h3Index.h"
+#include "test.h"
+
+SUITE(h3ToCenterChild) {
```
It wouldn't hurt to add an exhaustive test here - something like (pseudocode):
```
for res in 0 to 14:
  for childRes in res + 1 to 15:
    h3Index = indexAtRes(res)
    centerChild = h3ToCenterChild(h3Index, childRes)
    // some property-based tests
    t_assert(h3GetResolution(centerChild) == childRes)
    t_assert(h3GetParent(centerChild, res) == h3Index)
```

isaacbrodsky replied:
+1, I'd like to see a more exhaustive test for this function

zachasme replied:
Good suggestion. I've also moved here the check that the child should equal indexed centroid of parent at child resolution. Are there more properties that asserted here?

isaacbrodsky replied:
Only other check I can think to include is that the returned index passes `h3IsValid`

nrabinowitz replied:
The assertions here look good to me 👍

nrabinowitz commented on src/apps/testapps/testPentagonsAtRes.c:
```diff
@@ -0,0 +1,46 @@
+
+            for (int i = 0; i < PADDED_COUNT; i++) {
+                H3Index h = indexes[i];
+                if (H3_EXPORT(h3IsValid)(h)) {
```
Nit: This might make more sense as 
```
if (h) {
  t_assert(H3_EXPORT(h3IsValid(h)), "index should be valid");
  // ... etc
}
```
This might help differentiate errors where we got an invalid index from cases where the array position wasn't filled.

nrabinowitz commented on src/apps/testapps/testPentagonsAtRes.c:
```diff
@@ -0,0 +1,46 @@
+            int numFound = 0;
+
+            for (int i = 0; i < PADDED_COUNT; i++) {
+                H3Index h = indexes[i];
```
Nit: We're not very consistent here, but I think we generally use `h3` or `h3Index` as a conventional variable name.

nrabinowitz commented on src/h3lib/lib/h3Index.c:
```diff
@@ -214,6 +214,31 @@ void H3_EXPORT(h3ToChildren)(H3Index h, int childRes, H3Index* children) {
+ *
+ * @return H3Index of the center child, or 0 if you actually asked for a parent
+ */
+H3Index H3_EXPORT(h3ToCenterChild)(H3Index parentH, int childRes) {
```
Nit: I'd prefer just `parent` or `h` to `parentH`. Similarly, below, I'd just go with `child` or `childIndex` - we don't use `-H` as a suffix anywhere else.

nrabinowitz commented on src/h3lib/lib/h3Index.c:
```diff
@@ -214,6 +214,31 @@ void H3_EXPORT(h3ToChildren)(H3Index h, int childRes, H3Index* children) {
+        return H3_INVALID_INDEX;
+    } else if (childRes == parentRes) {
+        return parentH;
+    } else if (childRes < 0 || childRes > MAX_H3_RES) {
```
Nit 1: `childRes < 0` should not be possible if we've checked `childRes < parentRes`.
Nit 2: I'd combine this branch with the `if (childRes < parentRes)` branch above to consolidate validation.

More substantively - we need a similar check in `h3ToChildren` and `uncompact`, but it doesn't look like we check `MAX_H3_RES` there. Could we add a static function like `static bool isValidChildRes(int parentRes, int childRes)` and use in all three functions?

nrabinowitz commented on src/h3lib/lib/h3Index.c:
```diff
@@ -860,6 +885,29 @@ void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+void H3_EXPORT(getPentagonsAtRes)(int res, H3Index* out) {
+    int i = 0;
+    for (int bc = 0; bc < NUM_BASE_CELLS; bc++) {
+        if (_isBaseCellPentagon(bc)) {
```
Would it be easier just hard-coding the indexes of the 12 base cells here? I believe they're `4, 14, 24, 38, 49, 58, 63, 72, 83, 97, 107, 117` - we could add a constant in `baseCells.h`. OTOH, I doubt performance is an issue here, so 🤷‍♂

zachasme replied:
I did consider this but felt bad about adding a hard-coded list in addition to `BaseCellData` which already contains the information. What would you prefer?

nrabinowitz replied:
Probably best to leave it as is - still constant time and probably not worth over-optimizing on performance for this function. So nvm :).

nrabinowitz commented on src/h3lib/lib/h3Index.c:
```diff
@@ -860,6 +885,29 @@ void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+    int i = 0;
+    for (int bc = 0; bc < NUM_BASE_CELLS; bc++) {
+        if (_isBaseCellPentagon(bc)) {
+            H3Index pentagonH;
```
Nit: As above, prefer `pentagon` to `pentagonH`

nrabinowitz commented:
> What are your thoughts on naming?
> 
> The `atRes` suffix is not used anywhere else, perhaps `getPentagonalIndexes`/`pentgonalIndexCount` would be better?

We also don't use `pentagonal` anywhere :). I don't mind `atRes` but could see just `getPentagons` or `getPentagonIndexes`. No strong preference.

isaacbrodsky commented on docs/api/hierarchy.md:
````diff
@@ -26,6 +26,22 @@ int maxH3ToChildrenSize(H3Index h, int childRes);
+H3Index h3ToCenterChild(H3Index h, int childRes);
+```
+
+Returns the center child (finer) index containing `h` at resolution `childRes`.
````
"contained by `h`"?

isaacbrodsky commented on docs/api/misc.md:
````diff
@@ -73,3 +73,19 @@ int res0IndexCount();
+void getPentagonsAtRes(int res, H3Index *out);
+```
+
+All the pentagon **H3** indexes at specified resolution.
````
nit: "at the specified resolution."

Please add a note that `out` must be of size `pentagonIndexCount()`

isaacbrodsky commented on docs/api/misc.md:
````diff
@@ -73,3 +73,19 @@ int res0IndexCount();
+int pentagonIndexCount();
+```
+
+Number of pentagon **H3** indexes.
````
"indexes per resolution."

isaacbrodsky commented on src/apps/testapps/testH3ToCenterChild.c:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017-2019 Uber Technologies, Inc.
```
nit: could mark it as "Copyright 2019 Uber Technologies, Inc."

isaacbrodsky commented on src/apps/testapps/testPentagonsAtRes.c:
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2017-2019 Uber Technologies, Inc.
```
Same comment about copyright header

isaacbrodsky commented on src/apps/testapps/testPentagonsAtRes.c:
```diff
@@ -0,0 +1,46 @@
+                    t_assert(H3_EXPORT(h3IsPentagon(h)), "only pentagons");
+                    t_assert(H3_EXPORT(h3GetResolution(h)) == res,
+                             "only correct resolution");
+                    numFound++;
```
Can we check that the found pentagon indexes are all unique?

isaacbrodsky commented on src/h3lib/lib/h3Index.c:
```diff
@@ -860,6 +885,29 @@ void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+ *
+ * @return int count of pentagon indexes
+ */
+int H3_EXPORT(pentagonIndexCount)() { return 12; }
```
Nit: Move `#define NUM_PENTAGONS 12` from `generateNumHexagons.c` to `constants.h` and reuse it here?

isaacbrodsky commented on src/h3lib/lib/h3Index.c:
```diff
@@ -860,6 +885,29 @@ void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+/**
+ * Generates all pentagons at the specified resolution
+ *
+ * @param res The resolution to produce pentagons at
```
Please add a comment for `out` indicating how large the array should be

nrabinowitz commented on src/apps/testapps/testH3ToCenterChild.c:
```diff
@@ -0,0 +1,47 @@
+
+    TEST(invalidInputs) {
+        t_assert(H3_EXPORT(h3ToCenterChild)(hex8, 6) == 0,
+                 "Lower resolution fails");
```
Nit: We try to avoid "lower"/"higher" as potentially confusing - when referring to resolution, please use "finer" or "coarser".

zachasme commented:
Thank you for all the feedback! I believe I've addressed it all. Please let me know if I've missed something.

I've added `static bool _isValidChildRes(int, int)` which is now used in `h3ToChildren`, `maxH3ToChildrenSize`, `h3ToCenterChild`, `uncompact` and `maxUncompactSize` as per your suggestion.

nrabinowitz commented on docs/api/misc.md:
````diff
@@ -73,3 +74,20 @@ int res0IndexCount();
+int pentagonIndexCount();
+```
+
+Number of pentagon **H3** indexes per resolution.
````
Nit: Might be worth noting that this is always 12, but provided as a convenience. That particular number is fairly fundamental to the grid system, extremely unlikely to change, and included in the docs, so we probably don't need to pretend it's an implementation detail.

nrabinowitz commented on src/apps/testapps/testCompact.c:
```diff
@@ -180,6 +184,15 @@ SUITE(compact) {
+        }
+        uncompactResult = H3_EXPORT(uncompact)(
+            someHexagons, numHex, uncompressed, numHex * 7, MAX_H3_RES + 1);
+        fprintf(stderr, "@@\n%d\n", uncompactResult);
```
Please remove the `fprint` statement.

isaacbrodsky commented on src/h3lib/lib/h3Index.c:
```diff
@@ -145,6 +145,21 @@ H3Index H3_EXPORT(h3ToParent)(H3Index h, int parentRes) {
 }
 
+/**
+ * Determines whether a one resolution is a valid child resolution of another
```
I think "a" should be removed here

nrabinowitz commented on src/apps/testapps/testPentagonIndexes.c:
```diff
@@ -0,0 +1,59 @@
+                    int indexSeen = 0;
+                    for (int j = i + 1; j < PADDED_COUNT; j++) {
+                        if (h3Indexes[j] == h3Index) {
+                            indexSeen++;
```
Nit: You could also drop `indexSeen` and simply have `t_assert(false, "index should be seen only once")` if found

zachasme commented on src/apps/testapps/testCompact.c:
```diff
@@ -180,6 +184,15 @@ SUITE(compact) {
+        }
+        uncompactResult = H3_EXPORT(uncompact)(
+            someHexagons, numHex, uncompressed, numHex * 7, MAX_H3_RES + 1);
+        fprintf(stderr, "@@\n%d\n", uncompactResult);
```
Oops :blush:

isaacbrodsky commented on src/h3lib/lib/h3Index.c:
```diff
@@ -145,6 +145,21 @@ H3Index H3_EXPORT(h3ToParent)(H3Index h, int parentRes) {
 }
 
+/**
+ * Determines whether one resolution is a valid child resolution of another
```
nit: could also note that child being the same as parent is considered valid.
